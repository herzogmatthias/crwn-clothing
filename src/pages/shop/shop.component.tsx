import * as React from "react";
import CollectionsOverview from "../../components/collections-overview/collections-overview.component";
import { Route, RouteComponentProps } from "react-router-dom";
import CollectionPage from "../collection/collection.component";
import {
  firestore,
  convertCollectionsSnapshotToMap,
} from "../../firebase/firebase.utils";
import { updateCollections } from "../../redux/shop/shop.actions";
import { connect, ConnectedProps } from "react-redux";

type IShopPageProps = ConnectedProps<typeof connector> & RouteComponentProps;

function ShopPage({ match, updateCollections }: IShopPageProps) {
  let unsubscripeFromSnapshot = null;
  React.useEffect(() => {
    const collectionRef = firestore.collection("collections");
    collectionRef.onSnapshot(async (snapshot) => {
      const collectionsMap = convertCollectionsSnapshotToMap(snapshot);
      updateCollections(collectionsMap);
    });
    return () => {};
  }, []);

  return (
    <div className="shop-page">
      <Route
        exact
        path={`${match.path}`}
        component={CollectionsOverview}
      ></Route>
      <Route
        path={`${match.path}/:categoryId`}
        component={CollectionPage}
      ></Route>
    </div>
  );
}

const mapDispatchToProps = {
  updateCollections: updateCollections,
};

const connector = connect(null, mapDispatchToProps);
export default connect(null, mapDispatchToProps)(ShopPage);
