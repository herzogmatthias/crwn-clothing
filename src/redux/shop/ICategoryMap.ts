import { ICategory } from "./ICategory";

export interface ICategoryMap {
  [key: string]: ICategory;
}
