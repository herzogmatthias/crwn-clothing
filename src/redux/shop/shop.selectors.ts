import { RootState } from "../root-reducer";
import { createSelector } from "reselect";

const selectShop = (state: RootState) => state.shop;

export const selectShopCollections = createSelector(
  [selectShop],
  (shop) => shop.collections
);

export const selectCollectionsForPreview = createSelector(
  [selectShopCollections],
  (collections) =>
    collections
      ? Object.keys(collections).map((key: string) => collections[key])
      : []
);

export const selectCollection = (collectionUrlParam: string) =>
  createSelector([selectShopCollections], (collections) =>
    collections ? collections[collectionUrlParam] : null
  );
