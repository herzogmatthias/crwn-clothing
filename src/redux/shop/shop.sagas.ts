import { call, put, takeLatest, all } from "redux-saga/effects";
import { FETCH_COLLECTIONS_START } from "./shop.types";
import {
  firestore,
  convertCollectionsSnapshotToMap,
} from "../../firebase/firebase.utils";
import {
  fetchCollectionsSuccess,
  fetchCollectionsFailure,
} from "./shop.actions";

export function* fetchCollectionAsync() {
  try {
    const collectionRef = firestore.collection("collections");
    const snapshot = yield collectionRef.get();
    const collectionsMap = yield call(
      convertCollectionsSnapshotToMap,
      snapshot
    );
    yield put(fetchCollectionsSuccess(collectionsMap));
  } catch (error) {
    yield put(fetchCollectionsFailure(error.message));
  }
}

export function* onfetchCollectionsStart() {
  console.log("start fetching");
  yield takeLatest(FETCH_COLLECTIONS_START, fetchCollectionAsync);
}

export function* shopSagas() {
  yield all([call(onfetchCollectionsStart)]);
}
