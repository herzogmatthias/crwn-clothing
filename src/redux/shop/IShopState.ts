import { ICategoryMap } from "./ICategoryMap";

export interface IShopState {
  collections: ICategoryMap | null;
  isFetching: boolean;
  errorMessage: string;
}
