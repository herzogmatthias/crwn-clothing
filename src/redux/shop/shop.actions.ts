import { createAction, Dispatch } from "@reduxjs/toolkit";
import {
  FETCH_COLLECTIONS_START,
  FETCH_COLLECTIONS_SUCCESS,
  FETCH_COLLECTIONS_FAILURE,
} from "./shop.types";
import {
  convertCollectionsSnapshotToMap,
  firestore,
} from "../../firebase/firebase.utils";
import { withPayloadType } from "../genericActionPayloadType";
import { ICategoryMap } from "./ICategoryMap";

export const fetchCollectionsStart = createAction(FETCH_COLLECTIONS_START);
export const fetchCollectionsSuccess = createAction(
  FETCH_COLLECTIONS_SUCCESS,
  withPayloadType<ICategoryMap>()
);
export const fetchCollectionsFailure = createAction(
  FETCH_COLLECTIONS_FAILURE,
  withPayloadType<string>()
);
/**
 *  Dispatch async redux with redux-thunk
 *
 * @returns (dispatch: Dispatch) => void
 **/
/*
const fetchCollectionStartAsync = () => {
  return (dispatch: Dispatch) => {
    dispatch(fetchCollectionsStart());
    const collectionRef = firestore.collection("collections");

    collectionRef
      .get()
      .then((snapshot) => {
        const collectionsMap = convertCollectionsSnapshotToMap(snapshot);
        dispatch(fetchCollectionsSuccess(collectionsMap));
      })
      .catch((err) => {
        dispatch(fetchCollectionsFailure(err.message));
      });
  };
};
*/
