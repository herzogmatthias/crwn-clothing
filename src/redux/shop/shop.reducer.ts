import { IShopState } from "./IShopState";
import { createReducer, PayloadAction } from "@reduxjs/toolkit";
import { ICategoryMap } from "./ICategoryMap";
import {
  fetchCollectionsStart,
  fetchCollectionsSuccess,
  fetchCollectionsFailure,
} from "./shop.actions";

const initialState: IShopState = {
  collections: null,
  isFetching: false,
  errorMessage: "",
};

const shopReducer = createReducer(initialState, {
  [fetchCollectionsStart.type]: (state, action) => {
    state.isFetching = true;
  },
  [fetchCollectionsSuccess.type]: (
    state,
    action: PayloadAction<ICategoryMap>
  ) => {
    state.isFetching = false;
    state.collections = action.payload;
  },
  [fetchCollectionsFailure.type]: (state, action: PayloadAction<string>) => {
    state.isFetching = false;
    state.errorMessage = action.payload;
  },
});
export default shopReducer;
