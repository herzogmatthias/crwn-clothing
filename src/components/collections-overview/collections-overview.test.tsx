import React from "react";
import { ReactWrapper, mount } from "enzyme";
import { ICategoryMap } from "../../redux/shop/ICategoryMap";
import {
  default as ConnectedCollectionsOverview,
  CollectionsOverview,
} from "./collections-overview.component";
import configureStore, { MockStoreEnhanced } from "redux-mock-store";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";

describe("CONNECTED COLLECTIONS-OVERVIEW TESTS", () => {
  let wrapper: ReactWrapper;
  let mockStore = configureStore([]);
  let store: MockStoreEnhanced<unknown, {}>;
  const colllectionsMap: ICategoryMap = {
    hats: {
      id: "asdfsa33",
      items: [],
      routeName: "/hats",
      title: "Hats",
    },
    Mens: {
      id: "asdfsa34",
      items: [],
      routeName: "/mens",
      title: "Mens",
    },
  };
  beforeEach(() => {
    const mockState = {
      shop: {
        collections: colllectionsMap,
      },
    };

    store = mockStore(mockState);

    wrapper = mount(
      <BrowserRouter>
        <Provider store={store}>
          <ConnectedCollectionsOverview />
        </Provider>
      </BrowserRouter>
    );
  });
  test("render Collections-Overview Component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("check for props", () => {
    expect(
      wrapper.find(CollectionsOverview).prop("collections").length
    ).toEqual(Object.keys(colllectionsMap).length);
  });
});
