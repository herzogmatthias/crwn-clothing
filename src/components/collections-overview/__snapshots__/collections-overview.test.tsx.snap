// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CONNECTED COLLECTIONS-OVERVIEW TESTS render Collections-Overview Component 1`] = `
<BrowserRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "createHref": [Function],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Provider
      store={
        Object {
          "clearActions": [Function],
          "dispatch": [Function],
          "getActions": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
        }
      }
    >
      <Connect(CollectionsOverview)>
        <CollectionsOverview
          collections={
            Array [
              Object {
                "id": "asdfsa33",
                "items": Array [],
                "routeName": "/hats",
                "title": "Hats",
              },
              Object {
                "id": "asdfsa34",
                "items": Array [],
                "routeName": "/mens",
                "title": "Mens",
              },
            ]
          }
        >
          <styled.div>
            <div
              className="sc-fzoLsD iAaohI"
            >
              <CollectionPreview
                items={Array []}
                key="asdfsa33"
                routeName="/hats"
                title="Hats"
              >
                <styled.div>
                  <div
                    className="sc-Axmtr hTjEwh"
                  >
                    <styled.h1>
                      <h1
                        className="sc-AxmLO kBbsmV"
                      >
                        HATS
                      </h1>
                    </styled.h1>
                    <styled.div>
                      <div
                        className="sc-fzozJi kAIeAo"
                      />
                    </styled.div>
                  </div>
                </styled.div>
              </CollectionPreview>
              <CollectionPreview
                items={Array []}
                key="asdfsa34"
                routeName="/mens"
                title="Mens"
              >
                <styled.div>
                  <div
                    className="sc-Axmtr hTjEwh"
                  >
                    <styled.h1>
                      <h1
                        className="sc-AxmLO kBbsmV"
                      >
                        MENS
                      </h1>
                    </styled.h1>
                    <styled.div>
                      <div
                        className="sc-fzozJi kAIeAo"
                      />
                    </styled.div>
                  </div>
                </styled.div>
              </CollectionPreview>
            </div>
          </styled.div>
        </CollectionsOverview>
      </Connect(CollectionsOverview)>
    </Provider>
  </Router>
</BrowserRouter>
`;
